### DISCUSSION TITLE
Need help analysing log file. Leader election trigger, engine owner switch for no apparent reason.

### DISCUSSION THREAD

**Discussion by Dunge** [longhorn/longhorn#5926] [Q&A] [answered]
Infrastructure:
3 Ubuntu 22.04.2 LTS virtual machine on different physical hosts on a local network.
MicroK8s 1.26.4
Longhorn 1.4.1

From time to time (every couple of days), my cluster seems to break for 5-10 minutes. I see my services depending on redis having queries resulting in timeout, and redis seems to freeze because Longhorn is manipulating its storage.

I'm not suggesting Longhorn is causing the issue here, but since it's in the middle of it all I would like some help analyzing its logs in order to find out the actual cause of the error.

I believe the nodes hosted by microk8s are misbehaving, but I can't find any reason for them to be. Everything should be stable, no special application other than it runs on them, the host is stable, the network is stable, they have plenty of memory and disk space. I also have a Prometheus and Grafana installation allowing me to dig into thousands of different metrics, but I just can't find anything that would explain the issue.

From what I see of the logs below, it seems like there's a timeout happening on the K8S API and break the election process, print a bunch of Go error stack, and then start switching back and forth the engine and volume owners, which cause the downtimes. After that, it takes a couple of minutes before all depending services comes back online.

I tried to match the time with disk backup, it's not in the same period at all. I also looked for spikes in RAM or CPU usage on the nodes, can't find anything relevant.

If anyone could help me point out to anything I could look for more information it would be appreciated.

I censored my domain name replacing it with "mydomain" in the log file.

[Explore-logs-2023-05-12 13 24 39.zip](https://github.com/longhorn/longhorn/files/11466627/Explore-logs-2023-05-12.13.24.39.zip)

**Comment by Dunge** [comment]
Had another crash today.
Received a kubernetes event:

```
the disk default-disk-159a2b0f96e8c36d(/var/lib/longhorn/) on the node docker-1.mydomain.com has 11010048000 available, but requires reserved 5368709120, minimal 25% to schedule more replicas
```

Then 20sec later the node was marked as `NodeStatusUnknown` and all workload started to fail. A few seconds later it came back Ready and workloads cancelled their pod deletion.

I know my disk space is limited, but it still have lot of reserved space. That shouldn't be the reason to crash the whole thing!

**Reply by ChanYiLin** [reply]
For this case, maybe you can update the setting `Storage Over Provisioning Percentage` to 200 (%)
Or lower the `Storage Minimal Available Percentage` to 10 (%)

**Comment by ChanYiLin** [comment]
Also we have 1.4.2 released which improve some API server stability
You can also try to upgrade to 1.4.2, thanks!

**Reply by Dunge** [reply]
Hello @ChanYiLin , thanks for the reply!

`Storage Over Provisioning Percentage` is already set to 200%, I believe that's the default value. To be honest I can't understand what it mean based on the description.
But not having a replica being scheduled on a disk that is low on space is perfectly OK, this is not the problem I'm looking for. I just want to prevent complete crash to the system.

Thank you for the API server communication stability. I already upgraded to 1.4.2 last week. I believe my API server itself is the one that sometimes take a few seconds to reply for some reason I don't understand, and not the longhorn client side of things that was failing. But it's still causing major headaches when it starts failing.

My question is, can longhorn be the one that changes the node status to `NodeStatusUnkown` if it failed communication to the API or is longhorn reacting to the status change? Is there a setting I can increase the timeout to the API server before Longhorn starts taking action?

**Reply by ChanYiLin** [reply]
```
the disk default-disk-159a2b0f96e8c36d(/var/lib/longhorn/) on the node docker-1.mydomain.com has 11010048000 available, but requires reserved 5368709120, minimal 25% to schedule more replicas
```
With this message longhorn will change the node status to `NodeStatusUnkown `, it is not because of the API issue.
This means Longhorn believe that the disk space has hit the limit according to Longhorn calculation ([see document](https://longhorn.io/docs/1.4.2/volumes-and-nodes/node-space-usage/)) so the replica cannot be scheduled on this disk anymore.

ref:
- setting: https://longhorn.io/docs/1.4.2/references/settings/#storage-minimal-available-percentage
- how longhorn calculate the space: https://longhorn.io/docs/1.4.2/volumes-and-nodes/node-space-usage/

**Reply by Dunge** [reply]
But `NodeStatusUnknown` also add `NoSchedule` and `NoExecute` taints and start terminating all pods running on the node right? Isn't that a bit extreme for simply failing to schedule a replica? If I understand this correctly, most of the time that replica would not even be currently used (mounted) on that node directly because the workload would be elsewhere, and the replica is only used as a copy on standby.

Wouldn't it be better to just mark that volume as degraded and the replica as failing and if ever a pod wants to mount it on this node to fail the mount operation and force it to restart on a node that possesses replica of the volume? A bit like the behavior that happens if you set the desired replica amount to 1?

Because right now, completely killing the node send all workload to the others, who might not have enough ressources to accept it. It's much less desirable to kill the whole node than just failing a pod from attaching. Especially for the workload that is running perfectly fine on that node and do not require the volume, why would they get killed?

I'm just a novice user with very low knowledge in the storage domain so I know there's probably a good reason for this I'm not understanding, but I can't wrap my head around why it would be desirable to do that.

**Reply by Dunge** [reply]
@ChanYiLin Sorry to be persistent, but I really need to wrap this up and find what's wrong with my cluster.

I want to make sure we do not misunderstand each other, are you certain failing to schedule a replica will cause Longhorn to completely mark the node as down? This doesn't seem to be documented anywhere, there's no trace in the logs about this, and it seems completely illogical from my point of view to disrupt all healthy workload on a node because an unused replica suddenly grew a byte larger than the space available instead of just marking that replica as failed.


In any case, node went down again yesterday. The original cause remain a `NodeStatusUnkown` appearing out of nowhere. No message about replica failure this time (only had that once in many crashes). Node disk space remained at 69% used and I don't see any increase. Nothing that should be triggering that status.

Log is very similar to the original one posted above except without the GO stack errors and the leader election errors. Starts with a

```
time="2023-05-21T09:43:49Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"docker-3.mydomain.com\", UID:\"ea0df7c5-c412-466f-9c31-8b869cfa57d5\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"16827956\", FieldPath:\"\"}): type: 'Warning' reason: 'Ready' Kubernetes node docker-3.mydomain.com not ready: NodeStatusUnknown"
```

and back less than a second afterward:

```
time="2023-05-21T09:43:50Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"docker-3.mydomain.com\", UID:\"ea0df7c5-c412-466f-9c31-8b869cfa57d5\", APIVersion:\"longhorn.io/v1beta2\", ResourceVersion:\"16828254\", FieldPath:\"\"}): type: 'Normal' reason: 'Ready' Node docker-3.mydomain.com is ready"
```

But then I have a whole bunch of `Replica controller picked up` and `Instance Manager Controller docker-2.mydomain.com picked up instance-manager-e` and also `Operation cannot be fulfilled on nodes.longhorn.io \"docker-3.mydomain.com\": the object has been modified; please apply your changes to the latest version and try again"`

It went on like that for over 2 thousand lines, for 2 minutes until it stopped. Like if longhorn is juggling the same volumes back and forth over and over again every second for no reason. But it came back healthy after a while.

The important thing I'm looking for is the reason why the node get marked as `NodeStatusUnkown` in the first place.

**Reply by ChanYiLin** [reply, marked as answer]
Sorry for late reply
I just realize the `NodeStatusUnknown` you mentioned is not longhorn node CR but K8s node CR
Then you might need to check why the node became failed at the first place by
```
kubectl describe node 
```

You are right, Longhorn won't mark the node `NodeStatusUnknown` and migrate workload just because disk pressure
Replicas cannot be scheduled on that node is the only thing will happen.
Sorry for the misleading.

**Reply by Dunge** [reply]
Ok thanks for confirming, that makes more sense this way. I was about to switch longhorn for something else based on your comment to see if it was really the culprit, glad I won't have to.

Describing the node unfortunately say no additional information.

I also opened a ticket on the microk8 github with all my logs since it probably comes from there but unfortunately don't seems like I'll be having a reply
